src/compiler/chainrules.jl

```diff
@@ -18,6 +18,9 @@ such that if a suitable rule is defined later, the generated function will recom
 function has_chain_rrule(T)
   config_T, arg_Ts = Iterators.peel(T.parameters)
   configured_rrule_m = meta(Tuple{typeof(rrule), config_T, arg_Ts...})
+  
+  isnothing(configured_rrule_m) && return false, nothing  # too crude, surely
```

Comment:
Paging @mzgubic who I think knows about this story. What's the right way to handle this `nothing`?

Reply:
Bump... either @mzgubic or @oxinabox wrote this, I think.

Reply:
Hmm, I've forked your branch and commented out the line above, but can't reproduce this behaviour.
```julia
(opt_out_broadcasting) pkg> st
      Status `~/JuliaEnvs/test/opt_out_broadcasting/Project.toml`
  [082447d4] ChainRules v1.39.1
  [d360d2e6] ChainRulesCore v1.15.3 `dev/ChainRulesCore`
  [e88e6eb3] Zygote v0.6.41 `dev/Zygote`

julia> using Zygote

julia> using ChainRulesCore

julia> using ChainRules

julia> gradient(x -> sum(x .+ ones(2,2)), (1,2)) == ((2,2),)
true

```

Reply:
Right, I see no error in this configuration. ChainRules v1.39.1 is the tagged version, which does not have a rule matching this PR's `@opt_out`.

The error is when using this PR with https://github.com/JuliaDiff/ChainRules.jl/pull/644, and still happens for me.

Reply:
Ah, wow, missed the context entirely. In any case, I think what is happening is:
```julia
julia> config_T = Zygote.ZygoteRuleConfig{Zygote.Context}

julia> arg_Ts = Any[typeof(Base.Broadcast.broadcasted), typeof(+), Tuple{Int64, Int64}, Matrix{Float64}]

julia> configured_rrule_m = IRTools.meta(Tuple{typeof(rrule), config_T, arg_Ts...})
# nothing, which (I think) is incorrect

# and the reason is the method error here
julia> rrule(Zygote.ZygoteRuleConfig(), Base.Broadcast.broadcasted, +, (1, 2), rand(2, 2))
ERROR: MethodError: rrule(::Zygote.ZygoteRuleConfig{Zygote.Context}, ::typeof(Base.Broadcast.broadcasted), ::typeof(+), ::Tuple{Int64, Int64}, ::Matrix{Float64}) is ambiguous. Candidates:
  rrule(::RuleConfig{>:HasReverseMode}, ::typeof(Base.Broadcast.broadcasted), ::typeof(+), xs::Union{Number, Base.Broadcast.Broadcasted, AbstractArray{<:Number}, Tuple{Vararg{Number, var"#s3204"}} where var"#s3204"}...) in ChainRules at /Users/mzgubic/JuliaEnvs/test/opt_out_broadcasting/dev/ChainRules/src/rulesets/Base/broadcast.jl:157
  rrule(cfg::Zygote.ZygoteRuleConfig, ::typeof(Base.Broadcast.broadcasted), f::F, args::Vararg{Any, N}) where {F, N} in Zygote
Possible fix, define
  rrule(::Zygote.ZygoteRuleConfig, ::typeof(Base.Broadcast.broadcasted), ::typeof(+), ::Vararg{Union{Number, Base.Broadcast.Broadcasted, AbstractArray{<:Number}, Tuple{Vararg{Number, var"#s3204"}} where var"#s3204"}, N}) where N
Stacktrace:
 [1] top-level scope
   @ REPL[54]:1
```

Reply:
Thanks, I didn't realise there was an ambiguity. A nicer error would have helped me, but...

The ambiguity is this: I want to skip the "generic" CR rule, so I call this, and the new method is less specific than some other `rrule`s for specific operations like `+`:
```
julia> @macroexpand1 ChainRulesCore.@opt_out rrule(cfg::ZygoteRuleConfig, ::typeof(Broadcast.broadcasted), f::F, args::Vararg{Any,N}) where {F,N}
quote
    ((ChainRulesCore).no_rrule(cfg::ZygoteRuleConfig, ::typeof(Broadcast.broadcasted), f::F, args::Vararg{Any, N}) where {F, N}) = ChainRulesCore.nothing
    ((ChainRulesCore).rrule(cfg::ZygoteRuleConfig, ::typeof(Broadcast.broadcasted), f::F, args::Vararg{Any, N}) where {F, N}) = ChainRulesCore.nothing
end
```
Those more specific rules aren't so different from Zygote's, so it may not matter whose rule is called. But ideally this PR should probably opt out of all of them.

Reply:
No, this won't work. We can't opt out of e.g. the broadcasting rules NNlib defines. We need another plan here.